Add utility to open a window and run code in it for tests.

whiten opened the pull request:
Use this to start a test for navigator.locks, verifying that it works on
all browsers we care about.

When test-util.ts newTab() is called, a new browser window is opened,
with all console logs from it funnelled back to the main test process. A
module (usually that of the current test) may be loaded and all exported
symbols made available, allowing type-checked Typescript code to be
executed easily. The run() method on the resulting tab may then be
called to execute snippets of code, allowing for testing of interaction
between two windows at the same origin.

Example:

```
import {newTab} from './test-util';

test('storage', async () => {
  const tab = await newPage();
  await tab.run('window.localStorage.setItem("test", "modified")');
  expect(window.localStorage.getItem('test')).to.equal('modified');
}
```

whiten committed:
Add utility to open a window and run code in it for tests.

Use this to start a test for navigator.locks, verifying that it works on
all browsers we care about.

When test-util.ts newTab() is called, a new browser window is opened,
with all console logs from it funnelled back to the main test process. A
module (usually that of the current test) may be loaded and all exported
symbols made available, allowing type-checked Typescript code to be
executed easily. The run() method on the resulting tab may then be
called to execute snippets of code, allowing for testing of interaction
between two windows at the same origin.

Example:

import {newTab} from './test-util';

test('storage', async () => {
  const tab = await newPage();
  await tab.run('window.localStorage.setItem("test", "modified")');
  expect(window.localStorage.getItem('test')).to.equal('modified');
}

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+  // Mocha functions aren't available when importing from a new page.
+  if (typeof self.suiteSetup !== 'function') {
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    const noop = function () {};
```
[fyi] Another way to disable this test is to add a comment (I believe)

```
const noop = () => {
  // intentionally empty
};
```

arv replied:
Prefer arrow functions over function expressions. FunctionDeclaration is preferred at top level (and where required by semantics)

arv replied:
You can also do `const noop = () => undefined` which would have the same semantics as an empty function

whiten replied:
Thanks for the `undefined` tip, I didn't know that! Searched for other non-arrow functions and didn't find any.

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+type Response = {
+  tabId: string;
+  id: string;
+  result?: unknown;
```
optional unknown seems a bit strange. Just do unknown?

There is a new feature in TS4.4 which allows it to  distinguish between missing properties vs `| undefined`. We are not using that yet.

whiten replied:
If I omit the `?`, I get the below error later in the file. Anything else to try here?

```
Property 'result' is missing in type '{ tabId: any; id: any; }' but required in type 'Response'.`
test-util.ts[66, 3]: 'result' is declared here.
```

arv replied:
haha

OK TypeScript, you win this time.

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore No index signature of type 'string' was found on 'Console'.
+    console[level] = (...data: unknown[]) => {
+      window.opener.postMessage(
```
You can probably just call this synchronously across the frame barrier.

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+    // @ts-ignore No index signature of type 'string' was found on 'Console'.
+    console[level] = (...data: unknown[]) => {
+      window.opener.postMessage(
+        {tabId, id: 'log', result: {level, message: data.join(' ')}},
```
just `message: data` seems more accurate (and flexible)

whiten replied:
Yep, that's right. I had struggled to figure out how to then pass the resulting array back to message.log() and friends, but just got `...data` working. Thanks!

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  const AsyncFunction = Object.getPrototypeOf(async function () {}).constructor;
+  let module: typeof import('./test-util');
```
Interesting. I never had to import myself but I can see why you want this here.

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+  let module: typeof import('./test-util');
+  if (path !== undefined) {
+    try {
+      module = await import(window.origin + '/' + path);
```
You probably should do something like:

```ts
new URL('/' + path, location.href).toString();
```

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+  }
+
+  const handle = async (event: MessageEvent) => {
+    if (event.source === null || event.source != window.opener) {
```
Prefer triple equals at all times

whiten replied:
Fixed here and three other places in this PR. I'm going to follow up with an add of `eqeqeq` to .eslintrc.cjs to help avoid this.

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+    const response: Response = {tabId, id};
+    if (expr) {
+      try {
+        response.result = await AsyncFunction(
```
I see what you are doing here but I think you can make this simpler. Picking up the `AsyncFunction` function should not be needed. You can do this with plain old `Function`.

```ts
Function('module', `
  "use strict";
  Object.assign(self, module);
  return (async () => {
    ${expr}
  })();
`)(module);
```

whiten replied:
Good call, done. With an await on the resulting Function, I don't need to do one inside.

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+  if (tabCallbacks === undefined) {
+    return;
+  }
+  if (id == 'log') {
```
I think this would be simpler with direct calls instead of postMessage.

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+
+  // Open a new tab to a URL that doesn't run anything on load.
+  const tab = window.open(
+    document.location.origin + '/src/test-util.ts',
```
This loads the ts file as text. Probably not very meaningful. You can open `''` and you have communicate between these windows syncronously.

whiten replied:
It's the same domain though, allowing the use of `localStorage`! I didn't want to use the root page on this domain because that auto-runs the tests.

arv replied:
`''` is special

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+  tab.document.head.appendChild(script);
+
+  // Start initialization process, waiting for page to become live.
+  const interval = setInterval(() => {
```
The `script` element has an onload event. Not sure if that helps here though?

whiten replied:
I played with this a bunch, and was unable to prove to myself that it ever fired. I'm not confident it wouldn't work, but am giving up for now.

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+    tab.postMessage({id: 'init', tabId, path: filename}, window.origin);
+  }, 100);
+  await promise;
+  clearInterval(interval);
```
Love this pattern!

whiten replied:
Thanks!

arv commented on src/test-util.ts:
```diff
@@ -32,3 +34,195 @@ export function b(
+    run: async (expr: string) => {
+      const id = uuid();
+      const {promise, resolve, reject} = resolver<unknown>();
+      tabCallbacks[id] = {resolve, reject};
```
I would use a Map for a map

whiten replied:
Done, good suggestion.

arv commented on src/web-lock.test.ts:
```diff
@@ -0,0 +1,75 @@
+
+test('ifAvailable', async () => {
+  await navigator.locks.request('exlock', {mode: 'shared'}, async () => {
+    expect(await tab.run('await lockIfAvailable("exlock")')).to.equal(true);
```
I love how the exported bindings are also loaded into the new tab. Awesome work!

whiten replied:
Thank you! I went through several drafts where I had different snippets of pure JS in quoted strings, and it was just proving impossible to debug. I'm glad I figured out how to make this work.

arv reviewed (changes requested):
This is great!

I'm extra nitpicky since this is your first PR.

I think the main thing I would change is to make the communcation sync and not rely on postMessage. It would simplify a lot of the code I believe

For sync communications you can do:

```js
const w = window.open();
w.onload = () => {
  console.log(w.document.body.textContent);
  console.assert(w.opener === window);
};
```

Remember that all global (`globalThis`/`window`) bindings (variabels, functions etc) are visible. So if you set `globalThis.foo` in the opened window the opener can call this directly (it can be an async function etc).

aboodman commented:
One thing to think about here is the fact that it's possible for different tabs from the same origin to be in separate processes, with their own separate main threads. In the more common case, every tab from the same origin will be in the same process and on the same thread / message loop. It's worth asking whether we want to test that case.

The way you are opening the window here, the two tabs will be in same process and you can in fact do cross-frame synchronous calls as Erik mentions. If you want to cause the opened window to have its own process, then you have to open it with `rel=noopener`, but in that case the synchronous calls will stop working.

See https://jakearchibald.com/2016/performance-benefits-of-rel-noopener/ for more.

arv commented:
@aboodman Good point. Let's keep the postMessage goop for that reason.

whiten committed:
Respond to all code review feedback.

Also remove explicit `window.` in many places, and add a `withTimeout()`
utility to wrap a promise with a timeout. I used this for both new tab
initialization and calls to the new tab as working on tests is abysmal
if one makes a small error with the default 30s timeout, and this seemed
to help.a lot.

whiten committed:
Merge branch 'main' into weblock

whiten commented:
Okay, responded to all suggestions, sticking with `postMessage` to exercise cross-process cases as well. I accidentally introduced a merge commit into this, so let me know if you need me to rebase this again to excise that from the commit history here. Back to you, @arv!

arv commented on src/test-util.ts:
```diff
@@ -72,91 +69,103 @@ type Response = {
       }
     }
-    event.source.postMessage(response);
+    opener.postMessage(response, origin);
```
`opener` will not work if out of process I believe?

arv replied:
The old code should work on the other hand

arv commented on src/test-util.ts:
```diff
@@ -72,91 +69,103 @@ type Response = {
 
+function withTimeout<T = void>(promise: Promise<T>, ms: number) {
+  const timeout = new Promise((_resolve, reject) => {
+    setTimeout(() => reject(new Error(`Timeout after ${ms}ms.`)), ms);
```
you could clearTimeout when the other one is done but this is just test code so no big deal

arv commented on src/test-util.ts:
```diff
@@ -189,39 +199,39 @@ export async function newTab(filename?: string): Promise<Tab> {
-      tab.postMessage({id, expr}, window.origin);
-      const result = await promise;
-      delete callbacks[id];
+      tab.postMessage({id, expr}, origin);
```
I actually prefer `window.origin` over `origin`. I hardly every use `window.origin` so seeing it as identifier expression makes me look for the variable.

It is kind of a black art which properties of window are ok to access as identifier expressions so feel free to use your own judgement

arv reviewed (approved):
Feel free to deal with `opener` when/if it becomes an issue

arv commented:
I see some test failures as well.

whiten committed:
Extend timeouts within newPage().

whiten committed:
Merge branch 'main' into weblock

whiten committed:
Fix failure on WebKit on Linux.

I never figured out exactly what was going on here, but having the
import of testutil.ts happen in a handcrafted HTML file rather than
from a script element that's sideloaded in seems to resolve it reliably.

whiten committed:
Merge branch 'main' into weblock
